File: app/src/main/java/com/charlie/sawl/meal/MealTool.java
package com.charlie.sawl.meal;

import android.content.Context;
import com.charlie.sawl.R;
import com.charlie.sawl.tools.Preference;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

public class MealTool {
    public static final String MEAL_PREFERENCE_NAME = "MealData";
    public static final int TYPE_BREAKFAST = 1, TYPE_LUNCH = 2, TYPE_DINNER = 3;

    public static String getMealStringFormat(int year, int month, int day, int type) {
        month += 1;
        return year + "-" + month + "-" + day + "-" + type;
    }

    public static void saveMealData(Context mContext, String[] Calender, String[] Breakfast, String[] Lunch, String[] Dinner) {
        Preference mPref = new Preference(mContext, MEAL_PREFERENCE_NAME);
        SimpleDateFormat mFormat = new SimpleDateFormat("yyyy.MM.dd(E)", Locale.KOREA);

        for (int index = 0; index < Calender.length; index++) {
            try {
                Calendar mDate = Calendar.getInstance();
                mDate.setTime(mFormat.parse(Calender[index]));

                int year = mDate.get(Calendar.YEAR);
                int month = mDate.get(Calendar.MONTH) + 1;
                int day = mDate.get(Calendar.DAY_OF_MONTH);

                String mPrefBreakfastName = getMealStringFormat(year, month, day, TYPE_BREAKFAST);
                String mPrefLunchName = getMealStringFormat(year, month, day, TYPE_LUNCH);
                String mPrefDinnerName = getMealStringFormat(year, month, day, TYPE_DINNER);

                String mBreakfast = Breakfast[index];
                String mLunch = Lunch[index];
                String mDinner = Dinner[index];

                if (MealLibrary.isMealCheck(mBreakfast)) mBreakfast = "";
                if (MealLibrary.isMealCheck(mLunch)) mLunch = "";
                if (MealLibrary.isMealCheck(mDinner)) mDinner = "";

                mPref.putString(mPrefBreakfastName, mBreakfast);
                mPref.putString(mPrefLunchName, mLunch);
                mPref.putString(mPrefDinnerName, mDinner);

            } catch (ParseException e) {
                e.printStackTrace();
            }
        }
    }

    public static restoreMealDateClass restoreMealData(Context mContext, int year, int month, int day) {
        Preference mPref = new Preference(mContext, MEAL_PREFERENCE_NAME);
        SimpleDateFormat mCalenderFormat = new SimpleDateFormat("yyyy년 MM월 dd일", Locale.KOREA);
        SimpleDateFormat mDayOfWeekFormat = new SimpleDateFormat("E요일", Locale.KOREA);
        Calendar mDate = Calendar.getInstance();
        mDate.set(year, month, day);

        restoreMealDateClass mData = new restoreMealDateClass();

        String mPrefBreakfastName = getMealStringFormat(year, month + 1, day, TYPE_BREAKFAST);
        String mPrefLunchName = getMealStringFormat(year, month + 1, day, TYPE_LUNCH);
        String mPrefDinnerName = getMealStringFormat(year, month + 1, day, TYPE_DINNER);

        mData.Calender = mCalenderFormat.format(mDate.getTime());
        mData.DayOfTheWeek = mDayOfWeekFormat.format(mDate.getTime());
        mData.Breakfast = mPref.getString(mPrefBreakfastName, null);
        mData.Lunch = mPref.getString(mPrefLunchName, null);
        mData.Dinner = mPref.getString(mPrefDinnerName, null);

        if (mData.Breakfast == null && mData.Lunch == null && mData.Dinner == null) {
            mData.isBlankDay = true;
        }

        return mData;
    }

    static class restoreMealDateClass {
        String Calender, DayOfTheWeek, Breakfast, Lunch, Dinner;
        boolean isBlankDay = false;
    }

    public static boolean mStringCheck(String mString) {
        return mString == null || "".equals(mString) || " ".equals(mString);
    }

    public static todayMealData getTodayMeal(Context mContext) {
        Calendar mCalendar = Calendar.getInstance();
        int year = mCalendar.get(Calendar.YEAR);
        int month = mCalendar.get(Calendar.MONTH);
        int day = mCalendar.get(Calendar.DAY_OF_MONTH);

        restoreMealDateClass mData = MealTool.restoreMealData(mContext, year, month, day);
        todayMealData mReturnData = new todayMealData();

        if (!mData.isBlankDay) {
            int hour = mCalendar.get(Calendar.HOUR_OF_DAY);
            if (hour > 24 || hour <= 8) {
                mReturnData.title = mContext.getString(R.string.today_breakfast);
                mReturnData.info = (MealLibrary.isMealCheck(mData.Breakfast) ? mContext.getString(R.string.no_data_breakfast) : mData.Breakfast);
            } else if (hour < 14) {
                mReturnData.title = mContext.getString(R.string.today_lunch);
                mReturnData.info = (MealLibrary.isMealCheck(mData.Lunch) ? mContext.getString(R.string.no_data_lunch) : mData.Lunch);
            } else {
                mReturnData.title = mContext.getString(R.string.today_dinner);
                mReturnData.info = (MealLibrary.isMealCheck(mData.Dinner) ? mContext.getString(R.string.no_data_dinner) : mData.Dinner);
            }
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title);
            mReturnData.info = mContext.getString(R.string.no_data_message);
        }

        return mReturnData;
    }

    public static todayMealData todayBreakfast(Context mContext) {
        Calendar mCalendar = Calendar.getInstance();
        int year = mCalendar.get(Calendar.YEAR);
        int month = mCalendar.get(Calendar.MONTH);
        int day = mCalendar.get(Calendar.DAY_OF_MONTH);

        restoreMealDateClass mData = MealTool.restoreMealData(mContext, year, month, day);
        todayMealData mReturnData = new todayMealData();

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message);
            mReturnData.info = (MealLibrary.isMealCheck(mData.Breakfast) ? mContext.getString(R.string.no_data_message) : mData.Breakfast);
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title);
            mReturnData.info = mContext.getString(R.string.no_data_message);
        }

        return mReturnData;
    }

    public static todayMealData todayLunch(Context mContext) {
        Calendar mCalendar = Calendar.getInstance();
        int year = mCalendar.get(Calendar.YEAR);
        int month = mCalendar.get(Calendar.MONTH);
        int day = mCalendar.get(Calendar.DAY_OF_MONTH);

        restoreMealDateClass mData = MealTool.restoreMealData(mContext, year, month, day);
        todayMealData mReturnData = new todayMealData();

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message);
            mReturnData.info = (MealLibrary.isMealCheck(mData.Lunch) ? mContext.getString(R.string.no_data_message) : mData.Lunch);
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title);
            mReturnData.info = mContext.getString(R.string.no_data_message);
        }

        return mReturnData;
    }

    public static todayMealData todayDinner(Context mContext) {
        Calendar mCalendar = Calendar.getInstance();
        int year = mCalendar.get(Calendar.YEAR);
        int month = mCalendar.get(Calendar.MONTH);
        int day = mCalendar.get(Calendar.DAY_OF_MONTH);

        restoreMealDateClass mData = MealTool.restoreMealData(mContext, year, month, day);
        todayMealData mReturnData = new todayMealData();

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message);
            mReturnData.info = (MealLibrary.isMealCheck(mData.Dinner) ? mContext.getString(R.string.no_data_message) : mData.Dinner);
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title);
            mReturnData.info = mContext.getString(R.string.no_data_message);
        }

        return mReturnData;
    }

    public static class todayMealData {
        public String title, info;
    }
}

File: app/src/main/kotlin/com/charlie/sawl/meal/MealTool.kt
package com.charlie.sawl.meal

import android.content.Context
import com.charlie.sawl.R
import com.charlie.sawl.tools.Preference
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

object MealTool {
    const val MEAL_PREFERENCE_NAME = "MealData"
    const val TYPE_BREAKFAST = 1
    const val TYPE_LUNCH = 2
    const val TYPE_DINNER = 3

    fun getMealStringFormat(year: Int, month: Int, day: Int, type: Int): String {
        var month = month
        month += 1
        return "$year-$month-$day-$type"
    }

    fun saveMealData(mContext: Context, Calender: Array<String?>, Breakfast: Array<String?>, Lunch: Array<String?>, Dinner: Array<String?>) {
        val mPref = Preference(mContext, MEAL_PREFERENCE_NAME)
        val mFormat = SimpleDateFormat("yyyy.MM.dd(E)", Locale.KOREA)

        for (index in Calender.indices) {
            try {
                val mDate = Calendar.getInstance()
                mDate.time = mFormat.parse(Calender[index]!!)!!

                val year = mDate.get(Calendar.YEAR)
                val month = mDate.get(Calendar.MONTH) + 1
                val day = mDate.get(Calendar.DAY_OF_MONTH)

                val mPrefBreakfastName = getMealStringFormat(year, month, day, TYPE_BREAKFAST)
                val mPrefLunchName = getMealStringFormat(year, month, day, TYPE_LUNCH)
                val mPrefDinnerName = getMealStringFormat(year, month, day, TYPE_DINNER)

                var mBreakfast = Breakfast[index]!!.trim()
                var mLunch = Lunch[index]!!.trim()
                var mDinner = Dinner[index]!!.trim()

                if (MealLibrary.isMealCheck(mBreakfast)) mBreakfast = ""
                if (MealLibrary.isMealCheck(mLunch)) mLunch = ""
                if (MealLibrary.isMealCheck(mDinner)) mDinner = ""

                mPref.putString(mPrefBreakfastName, mBreakfast)
                mPref.putString(mPrefLunchName, mLunch)
                mPref.putString(mPrefDinnerName, mDinner)

            } catch (e: ParseException) {
                e.printStackTrace()
            }
        }
    }

    internal fun restoreMealData(mContext: Context, year: Int, month: Int, day: Int): RestoreMealDateClass {
        val mPref = Preference(mContext, MEAL_PREFERENCE_NAME)
        val mCalenderFormat = SimpleDateFormat("yyyy년 MM월 dd일", Locale.KOREA)
        val mDayOfWeekFormat = SimpleDateFormat("E요일", Locale.KOREA)
        val mDate = Calendar.getInstance()
        mDate.set(year, month, day)

        val mData = RestoreMealDateClass()

        val mPrefBreakfastName = getMealStringFormat(year, month + 1, day, TYPE_BREAKFAST)
        val mPrefLunchName = getMealStringFormat(year, month + 1, day, TYPE_LUNCH)
        val mPrefDinnerName = getMealStringFormat(year, month + 1, day, TYPE_DINNER)

        mData.calendar = mCalenderFormat.format(mDate.time)
        mData.dayOfTheWeek = mDayOfWeekFormat.format(mDate.time)
        mData.breakfast = mPref.getString(mPrefBreakfastName, null)
        mData.lunch = mPref.getString(mPrefLunchName, null)
        mData.dinner = mPref.getString(mPrefDinnerName, null)

        if (mData.breakfast == null && mData.lunch == null && mData.dinner == null) {
            mData.isBlankDay = true
        }

        return mData
    }

    internal class RestoreMealDateClass {
        var calendar: String? = null
        var dayOfTheWeek: String? = null
        var breakfast: String? = null
        var lunch: String? = null
        var dinner: String? = null
        var isBlankDay = false
    }

    fun mStringCheck(mString: String?): Boolean {
        return mString == null || "" == mString || " " == mString || "null" == mString
    }

    fun todayBreakfast(mContext: Context): TodayMealData {
        val mCalendar = Calendar.getInstance()
        val year = mCalendar.get(Calendar.YEAR)
        val month = mCalendar.get(Calendar.MONTH)
        val day = mCalendar.get(Calendar.DAY_OF_MONTH)

        val mData = restoreMealData(mContext, year, month, day)
        val mReturnData = TodayMealData()

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message)
            mReturnData.info = if (MealLibrary.isMealCheck(mData.breakfast)) mContext.getString(R.string.no_data_message) else mData.breakfast
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title)
            mReturnData.info = mContext.getString(R.string.no_data_message)
        }

        return mReturnData
    }

    fun todayLunch(mContext: Context): TodayMealData {
        val mCalendar = Calendar.getInstance()
        val year = mCalendar.get(Calendar.YEAR)
        val month = mCalendar.get(Calendar.MONTH)
        val day = mCalendar.get(Calendar.DAY_OF_MONTH)

        val mData = restoreMealData(mContext, year, month, day)
        val mReturnData = TodayMealData()

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message)
            mReturnData.info = if (MealLibrary.isMealCheck(mData.lunch)) mContext.getString(R.string.no_data_message) else mData.lunch
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title)
            mReturnData.info = mContext.getString(R.string.no_data_message)
        }

        return mReturnData
    }

    fun todayDinner(mContext: Context): TodayMealData {
        val mCalendar = Calendar.getInstance()
        val year = mCalendar.get(Calendar.YEAR)
        val month = mCalendar.get(Calendar.MONTH)
        val day = mCalendar.get(Calendar.DAY_OF_MONTH)

        val mData = restoreMealData(mContext, year, month, day)
        val mReturnData = TodayMealData()

        if (!mData.isBlankDay) {
            mReturnData.title = mContext.getString(R.string.no_message)
            mReturnData.info = if (MealLibrary.isMealCheck(mData.dinner)) mContext.getString(R.string.no_data_message) else mData.dinner
        } else {
            mReturnData.title = mContext.getString(R.string.no_data_title)
            mReturnData.info = mContext.getString(R.string.no_data_message)
        }

        return mReturnData
    }

    class TodayMealData {
        var title: String? = null
        var info: String? = null
    }
}